Microsoft.Toolkit.Services/PlatformSpecific/NETFramework/NetFrameworkPasswordManager.cs

```diff
@@ -0,0 +1,228 @@
+                        Flags = ncred.Flags,
+                        Persist = (CRED_PERSIST)ncred.Persist
+                    };
+                    return cred;
+                }
+                else
+                {
+                    throw new InvalidOperationException("Invalid CriticalHandle!");
+                }
+            }
+
+            // Perform any specific actions to release the handle in the ReleaseHandle method.
+            // Often, you need to use Pinvoke to make a call into the Win32 API to release the
+            // handle. In this case, however, we can use the Marshal class to release the unmanaged memory.
+            protected override bool ReleaseHandle()
+            {
+                // If the handle was set, free it. Return success.
+                if (!IsInvalid)
+                {
+                    // NOTE: We should also ZERO out the memory allocated to the handle, before free'ing it
+                    // so there are no traces of the sensitive data left in memory.
+                    CredFree(handle);
+
+                    // Mark the handle as invalid for future users.
+                    SetHandleAsInvalid();
+                    return true;
+                }
+
+                // Return false.
+                return false;
+            }
+        }
+
+        public void Store(string resource, PasswordCredential credential)
+        {
+            // Validations.
+            byte[] byteArray = Encoding.Unicode.GetBytes(credential.Password);
+            if (byteArray.Length > 512)
+            {
+                throw new ArgumentOutOfRangeException("The secret message has exceeded 512 bytes.");
+            }
+
+            // Go ahead with what we have are stuff it into the CredMan structures.
+            Credential cred = new Credential
+            {
+                TargetName = resource,
+                UserName = credential.UserName,
+                CredentialBlob = credential.Password,
+                CredentialBlobSize = (uint)Encoding.Unicode.GetBytes(credential.Password).Length,
+                AttributeCount = 0,
+                Attributes = IntPtr.Zero,
+                Comment = null,
+                TargetAlias = null,
+                Type = CRED_TYPE.GENERIC,
+                Persist = CRED_PERSIST.LOCAL_MACHINE
+            };
+            NativeCredential ncred = NativeCredential.GetNativeCredential(cred);
+
+            // Write the info into the CredMan storage.
+            bool written = CredWrite(ref ncred, 0);
+            int lastError = Marshal.GetLastWin32Error();
+            if (!written)
+            {
+                string message = string.Format("CredWrite failed with the error code {0}.", lastError);
+                throw new Exception(message);
+            }
+        }
+
+        public PasswordCredential Get(string key)
+        {
+            IntPtr nCredPtr;
+            PasswordCredential passCred = new PasswordCredential();
```

Comment:
this could be moved closer to where it is used and execution path cleaned up to make it clearer. e.g. `if (!CredRead(…)) { return null; } else { … }`

Reply:
Agree, move these down this method in the else where they are needed. The else is not necessary also